### ISSUE TITLE
Yolov5 dataset insights to improve mAP

### ISSUE THREAD

**Issue by Ownmarc** [ultralytics/yolov5#895]
## 🚀 Feature
Add ways for people to easily audit their dataset and labels using trained models. 

## Motivation

In my experience, the quickest way to gain mAP when starting model training on custom datasets will almost always be to review the labels for mistakes and correct them. I found that the models will, if provided enough data and a good validations set, be able to disagree with the labels and that you can use the trained model to find mistakes in the datasets. Correcting miss labelled data in the validation set will help get the true mAP of a model while correcting mistakes in the training dataset will help getting better results.

Here are a few ideas that I have implemented for myself to audit my datasets for labeling mistakes (they always happen, even with experimented labelers) : 

- **Objects cropping** Create as many folders as there are categories and clip all the labeled objects into small images (using the original filename in the name of the small image) and place them in their own category folder so the user can then visually inspect each folder for miss-labeled objects. It is really visual and it is easy to scroll through images. Most miss-labeled objects will standout when put side by side with only objects of the same category. 

- **Model biggest miss** Using the model to predict on our dataset and then sorting the images by worst results first. This would be an easy way for users to find out what images are difficult for the model to predict on. The user could then review the labels for those images first and if they are all correct, it would probably be a good idea to gather more images of this kind.

## Pitch

Adding active learning capabilities to this repo, making it easy to find ways to improve Yolov5 models.

## Alternatives

Open an other repo for that

**Comment by glenn-jocher** [comment]
@Ownmarc thanks for the suggestions! I agree a step in the direction of active learning / self labelling is the right direction for the future (global wheat showed this for example). Even on official datasets like VOC and COCO there are numerous labelling mistakes and omissions, and it would be awesome to automate the process of detecting and correcting these, and also to help facilitate the automatic scraping and autolabelling of new data as training progresses and overfitting nears.

There are a few tools (in this repo and elsewhere) already available that partially satisfy some of these tasks, the challenge is in execution, as always. I can help here by pointing out the repo's current capabilities, and we can use this thread to brainstorm next steps. 

- **Autolabelling tools.** Autolabelling is available through detect.py in batch-size 1, and also in test.py at any batch size using the --save-txt argument. For example to autolabel people in images I would use the best model with TTA:
```bash
python detect.py --source data/to/label/ --weights yolov5x.pt --save-txt --augment
```
This will create a folder of new labels text files (in the correct format). You can turn around and train with these labels with no other steps required.

- **Box cropping tools.**  detect.py offers an available second stage pass through a classification model, if you have one available, to reduce FPs, during which it crops each detected box and passes it to the classifier here:
https://github.com/ultralytics/yolov5/blob/d3f9bf2bb7932bcd696d3e062031c20ec0adda6c/detect.py#L80-L83. **Mass cropping of an entire dataset** is also available in the dataloader. If you have a detection dataset (like VOC or COCO), and you want to convert it into a classification dataset (like imagenet), with box contents cropped out and placed into class-labelled folders, you can do this here:
https://github.com/ultralytics/yolov5/blob/d3f9bf2bb7932bcd696d3e062031c20ec0adda6c/utils/datasets.py#L437-L454

- **Active learning tools.** Do not exist yet, but I agree with your ideas. I think that if a partially trained model (say 50% trained) repeatedly disagrees with a label (in a consistent way), this is probably a good indicator of a labeling error. Most human label errors will probably be false negatives (an object exists, but the human labeler missed it). On the contrary, autolabel errors will probably more often be mis-classifications, FPs, or incorrect boxes.

**Comment by AJ-RR** [comment]
Hi, could you please help me with some insight into how to tune the box loss gain, cls loss gain, cls BCELoss and object loss gain in order to improve map performance on a custom dataset. (evolution is too time consuming for me). Thanks in advance!

**Comment by www7890** [comment]
Hi, I am working on this research filed recently. I build up an architecture of active learning smart labeling.

The main concept is to take the output of confidence from images as consideration. The basic assumption is that higher confidence scores means higher quality of pre-label.

And like @glenn-jocher has mentioned before, mis-classifications, FPs, or incorrect boxes occur, and somehow take more time for human effort to correct them. I haven't found out a perfect solution to overcome the drawbacks of smart labeling.

Although there are lots of references on auto-labeling for object detection, all of them still has the same basic assumption: trust the values model gives us.
